### PR TITLE
Not doing superstructures got complicated fast.

### PR DESCRIPTION
**Pull request by Dwooll94**
* Prevent outer intake and lift from colliding
* Retract all
* Mode-based controls (Cargo Vs. Hatch) controls and forced all subsystems to respond to mode change event.

### CODE REVIEW

File: src/main/java/org/usfirst/frc/team1806/robot/OI.java [hunk 1]
```diff
@@ -49,59 +45,111 @@ public void runCommands(){
...
+				if(dc.getButtonRB() && !wasChangeModeButton){
+					Robot.setGamePieceMode(Robot.GamePieceMode.HATCH_PANEL);
+				}
+
+				break;
 		}
-		else if(dc.getButtonLB()){
-			mCargoIntakeSubsystem.scoreCargo(CargoIntakeSubsystem.ScoringPower.MEDIUM);
+
+		if(dc.getButtonBack()){
```
Comment (Dwooll94) [inline src/main/java/org/usfirst/frc/team1806/robot/OI.java]: Should the operator also get a button for retract all?

File: src/main/java/org/usfirst/frc/team1806/robot/OI.java [hunk 2]
```diff
@@ -49,59 +45,111 @@ public void runCommands(){
...
+						mLiftSubsystem.goToSetpoint(LiftSubsystem.LiftPosition.ROCKET_CARGO_LOW);
+					}
+					if(dc.getButtonX()) {
+						mLiftSubsystem.goToSetpoint(LiftSubsystem.LiftPosition.ROCKET_CARGO_MID);
+					}
+					if(dc.getButtonY()) {
+						mLiftSubsystem.goToSetpoint(LiftSubsystem.LiftPosition.ROCKET_CARGO_HIGH);
+					}
+					if(dc.getButtonB()) {
+						mLiftSubsystem.goToSetpoint(LiftSubsystem.LiftPosition.BOTTOM_LIMIT);
```
Comment (Dwooll94) [inline src/main/java/org/usfirst/frc/team1806/robot/OI.java:111]: Bringing the lift to the bottom is probably more common than switching modes... should we put that on RB and make B the switch mode button?

**Review by Dwooll94 — COMMENTED** [review]
Still some decisions to be made

File: src/main/java/org/usfirst/frc/team1806/robot/OI.java (hunk 2 above)
Comment (Orion024) [reply src/main/java/org/usfirst/frc/team1806/robot/OI.java:111]: I think we probably want most lift setpoints in the same area on the controller if possible, and having an accessible button for the switch mode will be easier to remember. It's all up to Joel though.

File: src/main/java/org/usfirst/frc/team1806/robot/Constants.java [hunk 3]
```diff
@@ -131,4 +131,13 @@
...
     public final static double habDropAccelerationThreshold = 1.0; //g force
 
+
+    //Lift interference avoidence
+    public final static double kLiftWaitForExtendIntake = 0.25;
+    public final static int kMaxLiftHeightToNeedToExtendIntake = 1000;
+    public final static int kSafeLiftHeightOffsetToNotHitIntake = 500; // total of this and the line above will be the setpoint
+
+    //controls
+    public final static double kTiggerThreshold = .2;
```
Comment (Orion024) [inline src/main/java/org/usfirst/frc/team1806/robot/Constants.java]: Refactor Tigger to Trigger

File: src/main/java/org/usfirst/frc/team1806/robot/subsystems/LiftSubsystem.java [hunk 4]
```diff
@@ -148,22 +190,14 @@ public void onLoop(double timestamp) {
...
+					lastTimeStamp = timestamp;
 				}
             }
 
-            private void cubePositionLoop(){
-            	switch (mLiftPosition){
-					case BOTTOM_LIMIT:
-						return;
-					case TOP_LIMIT:
-						return;
```
Comment (Orion024) [inline src/main/java/org/usfirst/frc/team1806/robot/subsystems/LiftSubsystem.java:163]: How is switch, case, return different from switch, case, break?

Comment (Dwooll94) [reply src/main/java/org/usfirst/frc/team1806/robot/subsystems/LiftSubsystem.java:163]: return will end the entire function you're in, and gives you the option to return a value at that point, break just gets you out of the lowest switch, while loop, if block, etc. that the break is in.

File: src/main/java/org/usfirst/frc/team1806/robot/Robot.java [hunk 5]
```diff
@@ -243,5 +244,32 @@ private void runTeleOp(){
...
     }
+
+    /**
+     * Sets the global game piece mode, runs functions associated with mode change on subsystems.
+     * @param mode the wanted mode
+     */
+    public static synchronized void setGamePieceMode(GamePieceMode mode){
+        GamePieceMode = mode;
+
+        switch(mode){
```
Comment (frc1806) [inline src/main/java/org/usfirst/frc/team1806/robot/Robot.java:255]: I think it's much simpler to directly call the goToCargoMode and goToHatchMode on the subsytems they apply to, not through the subsystem manager. If there are really only ~3 subsystems that need the methods, why are they on the interface? I think it's overdesigned in this respect

**Review by frc1806 — APPROVED** [review]

**Pull request merged**